pkg/linters/container/rules.go

```diff
@@ -0,0 +1,216 @@
+package container
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/deckhouse/d8-lint/internal/storage"
+	"github.com/deckhouse/d8-lint/pkg/errors"
+)
+
+const defaultRegistry = "registry.example.com/deckhouse"
+
+func applyContainerRules(object storage.StoreObject) (result errors.LintRuleErrorsList) {
+	containers, err := object.GetContainers()
+	if err != nil {
+		return
+	}
+	initContainers, err := object.GetInitContainers()
+	if err != nil {
+		return
+	}
+	containers = append(initContainers, containers...)
+	if len(containers) == 0 {
+		return
+	}
+
+	result = errors.LintRuleErrorsList{}
+
+	result.Add(containerNameDuplicates(object, containers))
+	result.Add(containerEnvVariablesDuplicates(object, containers))
+	result.Add(containerImageDigestCheck(object, containers))
+	result.Add(containersImagePullPolicy(object, containers))
+
+	result.Add(containerStorageEphemeral(object, containers))
+	result.Add(containerSecurityContext(object, containers))
+	result.Add(containerPorts(object, containers))
+
+	return result
+}
+
+func containersImagePullPolicy(object storage.StoreObject, containers []v1.Container) *errors.LintRuleError {
+	ob := object.Unstructured
+	if ob.GetNamespace() == "d8-system" && ob.GetKind() == "Deployment" && ob.GetName() == "deckhouse" {
+		c := containers[0]
+		if c.ImagePullPolicy != "Always" {
+			// image pull policy must be Always,
+			// because changing d8-system/deckhouse-registry triggers restart deckhouse deployment
+			// d8-system/deckhouse-registry can contain invalid registry creds
+			// and restarting deckhouse with invalid creads will break all static pods on masters
+			// and bashible
+			return errors.NewLintRuleError(
+				ID,
+				object.Identity()+"; container = "+c.Name,
+				c.Name,
+				c.ImagePullPolicy,
+				`Container imagePullPolicy should be unspecified or "Always"`,
+			)
+		}
+
+		return errors.EmptyRuleError
+	}
+
+	return containerImagePullPolicyIfNotPresent(object, containers)
+}
+
+func containerNameDuplicates(object storage.StoreObject, containers []v1.Container) *errors.LintRuleError {
+	names := make(map[string]struct{})
+	for i := range containers {
+		if _, ok := names[containers[i].Name]; ok {
+			return errors.NewLintRuleError(ID, object.Identity(), containers[i].Name, nil, "Duplicate container name")
+		}
+		names[containers[i].Name] = struct{}{}
+	}
+	return errors.EmptyRuleError
+}
+
+func containerEnvVariablesDuplicates(object storage.StoreObject, containers []v1.Container) *errors.LintRuleError {
+	for i := range containers {
```

Comment:
looks like we do not really need index here
```suggestion
	for i, container := range containers {
```

Reply:
containers here is a slice of structures, if we go through using range, then at each stage we will create a new variable, into which we will copy the structure. As a result, it is easier to read, but there are more memory costs and more CPU time for copying.